internal/rcm/api.go

```diff
@@ -0,0 +1,148 @@
+	"log"
+	"net"
+	"net/http"
+
+	"github.com/google/uuid"
+	"github.com/julienschmidt/httprouter"
+	"github.com/osbuild/osbuild-composer/internal/blueprint"
+	"github.com/osbuild/osbuild-composer/internal/store"
+	"github.com/osbuild/osbuild-composer/internal/target"
+)
+
+// API encapsulates RCM-specific API that is exposed over a separate TCP socket
+type API struct {
+	logger *log.Logger
+	store  *store.Store
+	router *httprouter.Router
+}
+
+// New creates new RCM API
+func New(logger *log.Logger, store *store.Store) *API {
+	api := &API{
+		logger: logger,
+		store:  store,
+		router: httprouter.New(),
+	}
+
+	api.router.RedirectTrailingSlash = false
+	api.router.RedirectFixedPath = false
+	api.router.MethodNotAllowed = http.HandlerFunc(methodNotAllowedHandler)
+	api.router.NotFound = http.HandlerFunc(notFoundHandler)
+
+	api.router.POST("/submit", api.submit)
+	api.router.GET("/status/:uuid", api.status)
+
+	return api
+}
+
+// Serve serves the RCM API over the provided listener socket
+func (api *API) Serve(listener net.Listener) error {
+	server := http.Server{Handler: api}
+
+	err := server.Serve(listener)
+	if err != nil && err != http.ErrServerClosed {
+		return err
+	}
+
+	return nil
+}
+
+// ServeHTTP logs the request, sets content-type, and forwards the request to appropriate handler
+func (api *API) ServeHTTP(writer http.ResponseWriter, request *http.Request) {
+	if api.logger != nil {
+		log.Println(request.Method, request.URL.Path)
+	}
+
+	writer.Header().Set("Content-Type", "application/json; charset=utf-8")
+	api.router.ServeHTTP(writer, request)
+}
+
+func methodNotAllowedHandler(writer http.ResponseWriter, request *http.Request) {
+	writer.WriteHeader(http.StatusMethodNotAllowed)
+}
+
+func notFoundHandler(writer http.ResponseWriter, request *http.Request) {
+	writer.WriteHeader(http.StatusNotFound)
+}
+
+func (api *API) submit(writer http.ResponseWriter, request *http.Request, _ httprouter.Params) {
+	// Check some basic HTTP parameters
+	contentType := request.Header["Content-Type"]
+	if len(contentType) != 1 || contentType[0] != "application/json" {
+		writer.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	// JSON structure expected from the client
+	type ComposeRequest struct {
+		Distribution  string   `json:"distribution"`
+		Architectures []string `json:"architectures"`
+		Repositories  []string `json:"repositories"`
+	}
+	// Parse and verify the structure
+	var composeRequest ComposeRequest
+	decoder := json.NewDecoder(request.Body)
+	decoder.DisallowUnknownFields()
+	err := decoder.Decode(&composeRequest)
+	if err != nil || composeRequest.Distribution == "" || len(composeRequest.Architectures) == 0 || len(composeRequest.Repositories) == 0 {
+		writer.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	// Push the requested compose to the store
+	composeUUID := uuid.New()
+	err = api.store.PushCompose(composeUUID, &blueprint.Blueprint{}, make(map[string]string), "test_arch", "test_output", &target.Target{})
```

Comment:
I"d make this somehow work, even when the other features are missing:
 - refuse requests that don't have exactly one architecture
 - pass the one architecture given instead of `test_arch`
 - pass `qcow2` instead of `test_output` (we should discuss if this should be configurable as well, but let's leave it like this for now)
 - set the target to a LocalTarget, the same way we do in weldr, this will be replaced by koji, but at least then we can test

Reply:
I cannot hardcode `qcow2`, because the package `test_distro` only works with `test_output`, otherwise I'll fix it.

Reply:
`api.store.PushCompose` creates the `LocalTarget` implicitly so there is no need to pass it

Reply:
Also note the "checksums" dictionary. It is supposed to be a checksum of repositories, but we want to pass the repositories as an argument, so what we really need is a new/modified argument completely.

Reply:
Yeah, you are right this all needs a bit of a rework. Feel free to leave this as is for the first version.